### PR TITLE
java_stub_template to use argument file instead of classpath jar

### PR DESCRIPTION
**Pull request by ittaiz**
### Description
java_stub_template to use argument file instead of classpath jar
This shaves 1s-1.5s when classpath is long.
Requires java > 8 so to properly add it in we'll probably have a toggle or use the toolchain or something
https://github.com/bazelbuild/bazel/issues/6354
### Motivation
performance

### CODE REVIEW

**Review by liucijus — CHANGES_REQUESTED** [review]
@ittaiz I think having an attribute as a toggle on the scala toolchain is the easiest way

**Comment by ittaiz** [comment]
@liucijus we're good to go. Build is green.
One thing I really don't like about the current test is that it doesn't really test anything. 
At maximum it tests there is such an attribute in the toolchain.
It doesn't check how it's used.
The main reason is because I'm not sure how to write non-flaky tests that compare perf.
Also the perf is felt when the classpath is huge so that is another problem.
I thought about maybe grepping through the bash script and see if it has True or False (after the substitutions that happen in starlark) but that feels wrong.

WDYT?

**Comment by ittaiz** [comment]
Btw, I think there's something flaky
https://buildkite.com/bazel/rules-scala-scala/builds/3951#0184ab62-2028-4d07-8a2f-d2048ed3a80f:
```

2022/11/24 20:48:36 Using unreleased version at commit 6d07ee5927283ecdab18822cc5a1a8a9a64d4074

Loading: 

Loading: 0 packages loaded

INFO: Build option --extra_toolchains has changed, discarding analysis cache.

Analyzing: target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider (0 packages loaded, 0 targets configured)

INFO: Analyzed target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider (0 packages loaded, 894 targets configured).

INFO: Found 1 target...

[0 / 2] [Prepa] BazelWorkspaceStatusAction stable-status.txt

ERROR: /workdir/test_expect_failure/missing_direct_deps/internal_deps/BUILD:117:14: scala @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider failed: (Exit 1): scalac failed: error executing command (from target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider) bazel-out/k8-opt-exec-2B5CBBC6/bin/src/java/io/bazel/rulesscala/scalac/scalac ... (remaining 1 argument skipped)

error: Target 'Unknown label of file bazel-out/k8-fastbuild/bin/test_expect_failure/missing_direct_deps/internal_deps/transitive_dependency_without_manifest.jar which came from @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency' is used but isn't explicitly declared, please add it to the deps.

You can use the following buildozer command:

buildozer 'add deps Unknown label of file bazel-out/k8-fastbuild/bin/test_expect_failure/missing_direct_deps/internal_deps/transitive_dependency_without_manifest.jar which came from @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency' //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider

one error found

Build failed

java.lang.RuntimeException: Build failed

	at io.bazel.rulesscala.scalac.ScalacWorker.compileScalaSources(ScalacWorker.java:290)

	at io.bazel.rulesscala.scalac.ScalacWorker.work(ScalacWorker.java:61)

	at io.bazel.rulesscala.worker.Worker.persistentWorkerMain(Worker.java:86)

	at io.bazel.rulesscala.worker.Worker.workerMain(Worker.java:39)

	at io.bazel.rulesscala.scalac.ScalacWorker.main(ScalacWorker.java:25)

Target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider failed to build

Use --verbose_failures to see the command lines of failed build steps.

INFO: Elapsed time: 1.029s, Critical Path: 0.93s

INFO: 4 processes: 2 internal, 2 worker.

FAILED: Build did NOT complete successfully

'bazel build //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider' should have logged "Unknown label of file .*transitive_dependency_without_manifest.jar which came from //test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency".

[0;31m Test "test_scala_library_expect_better_failure_message_on_missing_transitive_dependency_labels_from_other_jvm_rules" failed [0m (1 sec) [0m


```

File: scala/scala_toolchain.bzl [hunk 1]
```diff
@@ -137,6 +138,10 @@ scala_toolchain = rule(
             default = True,
             doc = "Enable writing of statsfile",
         ),
+        "use_argument_file_in_runner_for_improved_performance": attr.bool(
+            default = False,
+            doc = "Changes java binaries scripts (including tests) to use argument files and not classpath jars to improve performance, requires java > 8",
```
Comment (liucijus) [inline scala/scala_toolchain.bzl:143]: Did you mean JDK, with `java < 8`?

File: scala/scala_toolchain.bzl [hunk 2]
```diff
@@ -137,6 +138,10 @@ scala_toolchain = rule(
             default = True,
             doc = "Enable writing of statsfile",
         ),
+        "use_argument_file_in_runner_for_improved_performance": attr.bool(
```
Comment (liucijus) [inline scala/scala_toolchain.bzl:141]: I think shorter name is better: "use_argument_file_in_runner". Reasoning is already described in doc.

**Comment by liucijus** [comment]
> @liucijus we're good to go. Build is green. One thing I really don't like about the current test is that it doesn't really test anything. At maximum it tests there is such an attribute in the toolchain. It doesn't check how it's used. The main reason is because I'm not sure how to write non-flaky tests that compare perf. Also the perf is felt when the classpath is huge so that is another problem. I thought about maybe grepping through the bash script and see if it has True or False (after the substitutions that happen in starlark) but that feels wrong.
> 
> WDYT?

@ittaiz I don't think you should test perf with such tests. The goal of such test should be to understand if everything wires up correctly and does correct build. I don't think we have any easy way to do perf regression here. So we have to trust that you've done your own measurements.

**Review by liucijus — CHANGES_REQUESTED** [review]
In general LGTM, please shorten the name of the attribute

**Comment by liucijus** [comment]
> Btw, I think there's something flaky https://buildkite.com/bazel/rules-scala-scala/builds/3951#0184ab62-2028-4d07-8a2f-d2048ed3a80f:
> 
> ```
> 
> 2022/11/24 20:48:36 Using unreleased version at commit 6d07ee5927283ecdab18822cc5a1a8a9a64d4074
> 
> Loading: 
> 
> Loading: 0 packages loaded
> 
> INFO: Build option --extra_toolchains has changed, discarding analysis cache.
> 
> Analyzing: target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider (0 packages loaded, 0 targets configured)
> 
> INFO: Analyzed target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider (0 packages loaded, 894 targets configured).
> 
> INFO: Found 1 target...
> 
> [0 / 2] [Prepa] BazelWorkspaceStatusAction stable-status.txt
> 
> ERROR: /workdir/test_expect_failure/missing_direct_deps/internal_deps/BUILD:117:14: scala @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider failed: (Exit 1): scalac failed: error executing command (from target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider) bazel-out/k8-opt-exec-2B5CBBC6/bin/src/java/io/bazel/rulesscala/scalac/scalac ... (remaining 1 argument skipped)
> 
> error: Target 'Unknown label of file bazel-out/k8-fastbuild/bin/test_expect_failure/missing_direct_deps/internal_deps/transitive_dependency_without_manifest.jar which came from @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency' is used but isn't explicitly declared, please add it to the deps.
> 
> You can use the following buildozer command:
> 
> buildozer 'add deps Unknown label of file bazel-out/k8-fastbuild/bin/test_expect_failure/missing_direct_deps/internal_deps/transitive_dependency_without_manifest.jar which came from @//test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency' //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider
> 
> one error found
> 
> Build failed
> 
> java.lang.RuntimeException: Build failed
> 
> 	at io.bazel.rulesscala.scalac.ScalacWorker.compileScalaSources(ScalacWorker.java:290)
> 
> 	at io.bazel.rulesscala.scalac.ScalacWorker.work(ScalacWorker.java:61)
> 
> 	at io.bazel.rulesscala.worker.Worker.persistentWorkerMain(Worker.java:86)
> 
> 	at io.bazel.rulesscala.worker.Worker.workerMain(Worker.java:39)
> 
> 	at io.bazel.rulesscala.scalac.ScalacWorker.main(ScalacWorker.java:25)
> 
> Target //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider failed to build
> 
> Use --verbose_failures to see the command lines of failed build steps.
> 
> INFO: Elapsed time: 1.029s, Critical Path: 0.93s
> 
> INFO: 4 processes: 2 internal, 2 worker.
> 
> FAILED: Build did NOT complete successfully
> 
> 'bazel build //test_expect_failure/missing_direct_deps/internal_deps:unstamped_jar_dependent_on_some_java_provider' should have logged "Unknown label of file .*transitive_dependency_without_manifest.jar which came from //test_expect_failure/missing_direct_deps/internal_deps:unstamped_direct_java_provider_dependency".
> 
> �[0;31m Test "test_scala_library_expect_better_failure_message_on_missing_transitive_dependency_labels_from_other_jvm_rules" failed �[0m (1 sec) �[0m
> ```

I think tests got broken with  Bazel 6 changes in backwards incompatible way.

File: scala/scala_toolchain.bzl (hunk 1 above)
Comment (ittaiz) [reply scala/scala_toolchain.bzl:143]: I actually meant the JRE (e.g. the jvm).
Do you want me to change the phrasing? I can also write JDK though it's the JRE part of it that we care about. I don't remember if in the current state of affairs in bazel these concepts are bundled or not

File: scala/scala_toolchain.bzl (hunk 2 above)
Comment (ittaiz) [reply scala/scala_toolchain.bzl:141]: 👍🏽 will change

**Comment by ittaiz** [comment]
Re perf and test- my point is that if you would keep the attribute on the toolchain but stop using it and also of course not pass it to the binary runner script the test still passes.
I can grep through the script to see if the value (True/False) is passed to the script but that test is more brittle and white box.
Wdyt?

**Comment by ittaiz** [comment]
> I think grepping for the value is useful check

I'll change the test and assert on that

**Comment by ittaiz** [comment]
@liucijus done

**Comment by ittaiz** [comment]
I think the macos build is stuck for some reason. Maybe buildkite related. Running for 20m.
I tried canceling it from the UI but that didn't work.
Might need to timeout and  then rerun.
All other builds pased

**Comment by ittaiz** [comment]
@liucijus if you have a way to maybe stop and start the build that would be great.

**Comment by ittaiz** [comment]
pushed another commit since that build is stuck

**Review by liucijus — APPROVED** [review]

File: test/shell/test_scala_binary.sh [hunk 3]
```diff
@@ -18,5 +18,26 @@ test_scala_binary_expect_failure_on_missing_direct_deps_located_in_dependency_wh
   test_scala_library_expect_failure_on_missing_direct_deps ${dependency_target} ${test_target}
 }
 
+test_scala_binary_allows_opt_in_to_use_of_argument_file_in_runner_for_improved_performance() {
+
+  bazel run --extra_toolchains="//test/toolchains:use_argument_file_in_runner" //test/src/main/scala/scalarules/test/large_classpath:largeClasspath
+  grep "\"True\" == \"True\"" bazel-bin/test/src/main/scala/scalarules/test/large_classpath/largeClasspath
```
Comment (simuons) [inline test/shell/test_scala_binary.sh:24]: Hi @ittaiz, overall looks good to me. I just find this grep pattern non intuitive. Maybe template could render something like `test_runner_classpath_mode="argsfile"|"manifest"` do an if on that variable and then grep for `test_runner_classpath_mode="argsfile"` in test? wdyt?

Comment (ittaiz) [reply test/shell/test_scala_binary.sh:24]: sure. I can do that. I also didn't like the grep. Vaidas wanted the `use_argument_file` thing and I went with it all the way. Will do the change.
To make sure we're aligned. 
The new greps will be:
1. `"argsfile" == "argsfile"` 
2. `"manifest" == "argsfile"`
Right?

Comment (ittaiz) [reply test/shell/test_scala_binary.sh]: @simuons PTAL

**Review by simuons — APPROVED** [review]

**Comment by ittaiz** [comment]
./test_reproducibility.sh  is running for 2h. Probably flaky.

**Comment by liucijus** [comment]
Thanks, @ittaiz!

**Comment by ittaiz** [comment]
😊 sure thing. Thank you both for your review!

**Comment by ittaiz** [comment]
Oh, I just realized I didn't update the docs :(

**Pull request merged**